cmd/thanos/rule.go

```diff
@@ -594,19 +622,40 @@ func queryPrometheusInstant(ctx context.Context, logger log.Logger, addr, query
 	}
 	defer runutil.CloseWithLogOnErr(logger, resp.Body, "query body")
 
-	// Always try to decode a vector. Scalar rules won't work for now and arguably
-	// have no relevant use case.
+	// Decode only ResultType and load Result only as RawJson since we don't know
+	// structure of the Result yet.
 	var m struct {
 		Data struct {
-			Result model.Vector `json:"result"`
+			ResultType string          `json:"resultType"`
+			Result     json.RawMessage `json:"result"`
 		} `json:"data"`
 	}
-	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
+
+	if err = json.NewDecoder(resp.Body).Decode(&m); err != nil {
 		return nil, err
 	}
-	vec := make(promql.Vector, 0, len(m.Data.Result))
 
-	for _, e := range m.Data.Result {
+	var vectorResult model.Vector
+
+	// Decode the Result depending on the ResultType
+	// Currently only `vector` and `scalar` types are supported
+	switch m.Data.ResultType {
+	case promql.ValueTypeVector:
+		if err = json.Unmarshal(m.Data.Result, &vectorResult); err != nil {
+			return nil, err
+		}
+	case promql.ValueTypeScalar:
+		vectorResult, err = convertScalarJSONToVector(m.Data.Result)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, errors.Errorf("unknown response type: '%q'", m.Data.ResultType)
```

Comment:
We never use `q` so maybe let's stick to `%v` or `%s`? Does not matter much I think.